Remove fomantic transition module

silverwind opened the pull request:
Removes all dropdown and dimmer animations. Works everywhere as far as I can tell, but need to give this thorough testing. Removes around 70kb JS/CSS.

Note, I'm not 100% sure regarding the various callbacks, those will need more investigation, but it appears to work nonetheless.

Fixes: https://github.com/go-gitea/gitea/issues/15709

silverwind commented:
Has bugs with menu hiding in the navbar, wip.

silverwind commented:
Hiding bug fixed. Fomantic apparently never removes the `transition` class after transition, so I replicated that and that makes the hiding CSS work again.

delvh commented on web_src/js/modules/fomantic.js:
```diff
@@ -22,6 +22,41 @@ export function initGiteaFomantic() {
+    if (arg === 'show' || isIn) {
+      arg?.onStart?.(this);
+      ret = this.each((_, el) => {
+        $(el).removeClass('hidden');
+        $(el).addClass('visible');
```
I guess there's no way to use our own mechanism with `gt-hidden`, correct?

wxiaoguang replied:
No, they are different. Fomantic always use "hidden" internally.

silverwind replied:
No, other fomantic code relies on these classes.

delvh commented on web_src/js/modules/fomantic.js:
```diff
@@ -22,6 +22,41 @@ export function initGiteaFomantic() {
+    } else if (arg === 'hide' || isOut) {
+      arg?.onStart?.(this);
+      ret = this.each((_, el) => {
+        $(el).addClass('hidden');
+        $(el).removeClass('visible');
```
Same as above.

wxiaoguang commented:
It breaks the modal dialogs.

silverwind commented:
I see what you mean, the position is incorrect, but modal works otherwise. This was not an issue in an earlier iteration of this, checking...

<img width="1265" alt="Screenshot 2023-08-14 at 15 01 49" src="https://github.com/go-gitea/gitea/assets/115237/3083b344-5b76-4efb-b034-e37949e85d88">

silverwind commented:
Problem is the `display: flex !important;` inline style is missing, but interestingly, the [module](https://github.com/fomantic/Fomantic-UI/blob/develop/src/definitions/modules/transition.js) does not contain `flex` in the source.

silverwind commented:
Found it:

https://github.com/fomantic/Fomantic-UI/blob/de758dcb3b4b51657623d43893bc5351aadaf942/src/definitions/modules/transition.js#L236-L239

silverwind commented:
Modal should be fixed. This replacement function is likely not 100% equivalent to the removed module, but it appears to work well enough.

wxiaoguang commented:
TBH, I am not a fan of removing all animations.

According to modern web sites:

1. For dropdowns, it doesn't need to use animation in most cases
2. For fly-in and modal dialogs, in most cases, they use animations, I have checked Apple, Google


So I do not think it's right to remove the animations totally.

silverwind commented:
I like them nowhere and selectively allowing some would be a much more deeper dive into the hacky internals of Fomantic that I'm not interested in doing.

silverwind commented:
Checked how GitHub does it, they have a 300ms `animation` on the modal content, but not the backdrop, which shows instantly. Hide is instant on both backdrop and modal.

silverwind commented:
Modal now fades in. I also added a new CSS variable for the backdrop color and very slightly reduced it's opacity. I think it will be useful for themes.

<img width="619" alt="Screenshot 2023-08-15 at 15 21 30" src="https://github.com/go-gitea/gitea/assets/115237/230a4ee4-1638-4691-870c-f2a9c509a07e">
<img width="614" alt="Screenshot 2023-08-15 at 15 21 41" src="https://github.com/go-gitea/gitea/assets/115237/68320921-38c5-40ab-9041-b140b9d1b762">

silverwind commented:
One more tweak done: added box shadows to dropdowns so they stand out more against the background:

<img width="894" alt="Screenshot 2023-08-16 at 00 36 24" src="https://github.com/go-gitea/gitea/assets/115237/87645fd0-8235-4cda-9ab0-040815b9fdcc">

<img width="892" alt="Screenshot 2023-08-16 at 00 36 12" src="https://github.com/go-gitea/gitea/assets/115237/846fefdc-27a1-43cd-bee7-634d6d8dfa97">

<img width="253" alt="Screenshot 2023-08-16 at 00 26 25" src="https://github.com/go-gitea/gitea/assets/115237/27b79385-7a7d-4f93-8a50-449474a95342">
<img width="253" alt="Screenshot 2023-08-16 at 00 26 14" src="https://github.com/go-gitea/gitea/assets/115237/b23cea80-ac98-4184-83b0-5979395e38aa">

(subpixel misalignment in second screenshot is a fomantic bug)
also, fixed an issue with the border color changing unexpectedly on focus change.

wxiaoguang reviewed (approved):
Haven't tested, while it might be worth to try.
